### ISSUE TITLE
Compatibility issue with krb5 in canonicalisation of principal names

### ISSUE THREAD

**Issue by filipnavara** [SteveSyfuhs/Kerberos.NET#302]
**Describe the bug**

When krb5 sends the AS-REQ it sends the Principal name unqualified without the realm (eg. `user`). It expect the reply to come back with unqualified principal name. However, Kerberos.NET uses a qualified one (eg. `user@CORP2.IDENTITYINTERVENTION.COM`) in the reply.

AS-REQ:
```
6A81D93081D6A103020105A20302010AA31A3018300AA10402020096A2020400300AA10402020095A2020400A481AD3081AAA00703050000010010A111300FA003020101A10830061B0475736572A2201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA3333031A003020102A12A30281B066B72627467741B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA511180F32303232303730323131323130355AA7060204105D2886A81A301802011202011102011402011302011002011702011902011A
```
AS-REP:
```
6B8203533082034FA003020105A10302010BA2023000A3201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA430302EA003020101A12730251B237573657240434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA5820186618201823082017EA003020105A1201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA2333031A003020102A12A30281B066B72627467741B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA382011E3082011AA003020112A28201110482010DF3E88254343D4D4D1A2E43E91C4383EC14B32BFEE3603A2C742F3C04BFEE3609B63C15704EB116723C7EACCAF7743D6ECC59090B2AEF0BD4A204F91307F6F83D918E8446FFABFD33D29D6DF6FD3364BEF376F24B5441F0847B34F2D0E3949A72EE6610EF0C9E32A45CC63EB45319B7D013CCC445CA582129BD5005440F091F5822DB2845E0157F4E869BB61D4D3CF1AA756943D552BB0BD00201286108F1A814112EA513FE0F64FF2CAB884A4F19DDCF1868267FE64B87EB8CF504C38975671D68741919AED95F58D9F7BF5AA6CF4B38BE71F2A19DB23E077B96BF3F7F26D803104E41A9BEAB23E6308CBE277EE1853203C3F529EC40B234254988D04DCD31ED47204F62B21DD1BA7CC485CA3FA682015F3082015BA003020112A28201520482014E9EB2DB5F1D9A347A66A41A1351B62AB1638BAF0401EB529E85803E9FEB1A798503083FA4314304FBB34A76E535155AB7D8F420930994C4B6D0BD9E40D1F340FD996ACD180414E4531776FA5D48771BE8A1B58B08AF8CAD044508230EBC8DEF8EEAF206CF19D910E914D946D29C887A333D56CC333850DADA2921598238F44791B404C74B3EB7C51A9A2D7E9A376DEE71286C4E22140D3B62A9B04A25F4495DC7364EC70DB430ABDC7A2D172CB398F5D54F57F01F0E2841E37B595815F85C14D26C7B611D003386E95E9FEF141F925FB689555418FFB9613149D03708C5AB686845F29956F380C932B4D166869876212806B7B3B88FC0B1B01A7362F6F4E59154006E2C20E8A3172B32B0E7B386D885D03E15B55F67813688621A61973D2B7F2FFF063A8CB1AB0500F4AA0D3B4DBAFFDC6F494637D356A5D06883BC1C8B0A5ED4AFA82161FA1FFAE4B4A514533340
```

The name with the realm is constructed [here](https://github.com/dotnet/Kerberos.NET/blob/2b567a4cd64ca56de40c0d76ae25949e89025f21/Kerberos.NET/Entities/Krb/KrbKdcRep.cs#L221).

**To Reproduce**

Sample project available at https://github.com/filipnavara/KerberosLoopback.

.NET 7 Preview 6 (or nightly build) is required. Reproduces on Linux. Just run `dotnet run`. Note that it's possible to get this working in GitHub Codespaces but the .NET 7 daily builds have to be installed first.

**Expected behavior**
I would expect the KDC to be compatible with krb5 and the authentication to succeed.

**Screenshots**

<img width="938" alt="image" src="https://user-images.githubusercontent.com/1764393/176885366-ba416a82-93b0-4cf1-a5e9-9d19fb936787.png">
<img width="925" alt="image" src="https://user-images.githubusercontent.com/1764393/176885509-e8856dc0-ed50-4808-87bd-47dc251fe0d5.png">

**Comment by filipnavara** [comment]
On second though I may be misrepresenting the problem. I set `canonicalize = 1` in the krb5.conf file in an attempt to get krb5 to accept the responses at all. If I remove that the `CName` is set properly but krb5 rejects the response.

**Comment by filipnavara** [comment]
I turned off the canonicalization and reached error on a different request:

AS-REQ:
```
6C82041830820414A103020105A20302010CA382036B308203673082025CA103020101A28202530482024F6E82024B30820247A003020105A10302010EA20703050000000000A38201636182015F3082015BA003020105A1201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA2333031A003020102A12A30281B066B72627467741B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA381FC3081F9A003020112A281F10481EE96C05C8F604FAB2AC693D2C456AE500EF715CA8A80D7E418D6400BBB6FA10787947436C9280CB90DD20BF51EA421D9BB0710C7A460C4505C2E6C5F964EA3E044BEC4A6191E2E599257E09147B3E659101A645F7EC58E823D95A9BFE9984D1D91134A9B821569049DD5D4344AF2DB95F8D68859544031E1B0E9F2D29B8EF659773B0302FB50C24EC4795C37883CBCC117560528EAB8C786AA479D5970FD3F8261D646A8CA7D0FD5C6A1CE732449F7BC29FB885D1B5F447A8A74EC352063646042F0322A3FB001E86EF333E49AC534F7FF2A032377D8B07C2F4A5BDAF8EE357ACE64B3B72B8FAAF9CBE4A4FA076B46A481CA3081C7A003020112A281BF0481BCE9DD699510CBE4454272F9B6B4A163A527141E3D84D826BCCF935503C18A23C56B8B1821A2309BCD2E6962F8404C6C2DC5728CD064E5F458486F715FCB3E5F8FBF50596F4199414F61A5E03EA5833F1E39B2336C030BD845E3EF9382D5A3C53DC41B3EEF355B9972A128BBF0485AC70C7575F93B35B59F97570C773693048FC9CD4C072506DFF4EDF6F0F80F32C17D8C86E698BFA1D137B9888FEFC3F8B7262C382E86F2C56FEC3D8CFD9FFAD789940AF2911E4625C2DB4CBE6E124E30820103A10402020088A281FA0481F7A081F43081F1A1173015A003020110A10E040C4D6081BC20393087FAF8E358A281D53081D2A003020112A281CA0481C7E21845B091C9706F44195668E3F5C79AF389E5A48B34FBF0BF368DE66D952F2AD8FFA1F2B7EC4A57F5BFE7FA57067AA5C27F9613FD1737356BC838AB4058123EB4CD7B5C391D52567B814E5C5A48156F03E77AF292DF6FD5AAC8206564F28F98E9D013DEA2FD2BDF672998F1BDAA8A1A80EAE06574481DBDC66B0D2F846BB2C75D2D2E91A05817C6BBABCFEDB077B9CB8F8945F8A3F4AFED561C18115C78BE60D73D3868E1632C7EAE479E2433056458AB1E05EEC40A89492A59D2646F32B776B4EFB0633332F3A48198308195A00703050040810000A2201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA331302FA003020103A12830261B04485454501B1E636F7270322E6964656E74697479696E74657276656E74696F6E2E636F6DA511180F32303232303730313233313134355AA70602044B67A882A81A301802011202011102011402011302011002011702011902011A
```

AS-REP:
```
6D8203233082031FA003020105A10302010DA3201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA430302EA003020101A12730251B237573657240434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA582016F6182016B30820167A003020105A1201B1E434F5250322E4944454E54495459494E54455256454E54494F4E2E434F4DA231302FA003020102A12830261B04485454501B1E636F7270322E6964656E74697479696E74657276656E74696F6E2E636F6DA382010930820105A003020112A281FD0481FA1BE9FE0395A7A2966C9D150B24490317253726579EF51D0546B60928D8101B88926755CD3537167B8E8FBDDCB5C63E34E9FDB376BE33A9CF0A406DD5A026C239D1835CF22AAFA1095390D4A479ADFE91BECA84C95140AFA021CDE42EA409FACC373AEF79CA1F540FC24FC6E70EBC0B27287FEEC0292721611DE9E72F74E02EA5E72F096521DB196B0B8A91B760D148619830C19746D8335E23D52BABA11296440008CA724EC6A6850CF9887313563B7A45383D2491B0A55D57C7E9B3EFA45F607982BCA00BBA5D4A145AFC64EC909BF7184282B240C29F2307D241E4C4CDDDAA7845DA461624ABD18E1BBA56A2E64665B4C3C1CEF0C535492201A682014A30820146A003020112A282013D04820139696549E10FCCD2C85AB47B1673A20948EACFEDFF59DD93D3B329F8BD22D3FC965A6B3B81374CECD4AAB53E54990A51991DBB07B458FDA0AFD30FA4916FD25880A1E18E45F42FC3A7BC9C5DAF68C3CE42514ACE331ED8CFFAD2693B1B9502716F1566D0C197D772945A379D7A1A120C008E7CE4E95F76A6AEC6C14676B494E64B4A48341FB0B504745542AB6D839850CCBA36C9038E9DB542D171B0FEEB2420FFB8A6380CB48E2E5F2311677E43535397A1AE3111A14DD24535FB259C046616118A1749245D33D797093EFB354B70D74F28A990A85D26BE1C50323DB31ED24C37E1AC46E38FAD5D0F7E44D5BB192D01DBF6837020011A8BBF3DB1DC108752C46A568E8DBC607336E409194D1DE55C7D14B3D18CA1A039CEC2E3EE5957038DFBD4C62F1D1B82D4EDCDAD8DD4959EB2DEEB53123D1ACA7926EB2B
```

It fails the [check here](https://github.com/krb5/krb5/blob/34625d594c339a077899fa01fc4b5c331a1647d0/src/lib/krb5/krb/gc_via_tkt.c#L267):
```
 dec_rep->client: user\@CORP2.IDENTITYINTERVENTION.COM@CORP2.IDENTITYINTERVENTION.COM
 tkt->client: user@CORP2.IDENTITYINTERVENTION.COM
```

The problem likely happens earlier since `user@CORP2.IDENTITYINTERVENTION.COM` is already present in the CName of the request.

**Comment by filipnavara** [comment]
<img width="914" alt="image" src="https://user-images.githubusercontent.com/1764393/176902633-c6ef28e8-c981-4f3b-98be-dc74c49a46aa.png">
<img width="788" alt="image" src="https://user-images.githubusercontent.com/1764393/176902912-6b6a77e8-bbdb-45a5-86f1-cf24f796cdf4.png">

**Comment by filipnavara** [comment]
I'm no expert here but I am still wondering whether [this code](https://github.com/dotnet/Kerberos.NET/blob/2b567a4cd64ca56de40c0d76ae25949e89025f21/Kerberos.NET/Entities/Krb/KrbKdcRep.cs#L221) is supposed to append the realm. It passes the test if I remove it.

The Kerberos specification states the following:
>    crealm
      This field contains the name of the realm in which the client is
      registered and in which initial authentication took place.
>
>   cname
      This field contains the name part of the client's principal
      identifier.

That makes me feel that realm is supposed to be only in the first field and not in `cname`.

**Comment by SteveSyfuhs** [comment]
We had hit this issue in the past, except it was occurring in the AS- flow. As such we did fix it there, but I guess we never saw a krb5 server without the canonicalize bit set, so the fix never made it down to the generated service ticket. I'm a bit surprised the server is having an issue with this in particular because the caller should ultimately be in control of the canonicalize flag.

https://github.com/dotnet/Kerberos.NET/blob/aa4b19a43b89ff6e67df55da08199f4a5eae0c47/Kerberos.NET/Server/KdcAsReqMessageHandler.cs#L197-L200

In any case, I *think* the solution is to move the check to `GenerateServiceTicket`. Annoyingly `ServiceTicketRequest` doesn't carry the KdcOptions flags so that'll have to be extended.

**Comment by filipnavara** [comment]
In the AS-REQ in https://github.com/dotnet/Kerberos.NET/issues/302#issuecomment-1172340327 the CName is not present in the request body as far as I can see. So I don't think there's anything that can be propagated into `SamAccountName`.

(UPD: Eh, not sure the comment references the correct part of the traced requests. I can re-run the sample but there was definitely one place where `SamAccountName` was not set and the request didn't look like it contains CName)

**Comment by SteveSyfuhs** [comment]
No, it would need to be copied over from the TGT. Which I guess actually moves the fix up to `KdcTgsReqMessageHandler`

https://github.com/dotnet/Kerberos.NET/blob/2b567a4cd64ca56de40c0d76ae25949e89025f21/Kerberos.NET/Server/KdcTgsReqMessageHandler.cs#L186-L298

For context, by the time this method is called, the request has already been validated, user has already been found (`TransformClientIdentity`), and now all we're doing is dumping the result into a ticket for the caller.

It would have to be something like 

if (tgsReq.Body.KdcOptions & Canonicalize) {
request.SamAccountName = context.GetState<TgsState>(PaDataType.PA_TGS_REQ).DecryptedApReq.Ticket.CName
}

Though I'm somewhat wondering what CName looks like in this case. It Should be covered by the AS fix already.

**Comment by filipnavara** [comment]
Adding this to the place you pointed out makes it work:
```csharp
            if (tgsReq.Body.KdcOptions.HasFlag(KdcOptions.Canonicalize))
            {
                rst.SamAccountName = context.GetState<TgsState>(PaDataType.PA_TGS_REQ).DecryptedApReq.Ticket.CName.FullyQualifiedName;
            }
```

That said, I am still questioning whether appending the realm in in the `SamAccountName == null` case of `CreateCNameForTicket` is correct.

**Comment by SteveSyfuhs** [comment]
Ship it (assuming unit tests pass).

> That said, I am still questioning whether appending the realm in in the SamAccountName == null is correct.

It's not as per RFC, but for compatibility and interop reasons it does need to be there.

**Comment by filipnavara** [comment]
Cannot run the unit tests on my machine at the moment. I opened draft PR https://github.com/dotnet/Kerberos.NET/pull/303.

**Comment by SteveSyfuhs** [comment]
PR build will run them. Build will fail because of a package push issue, but that's expected for PRs from folks outside the org. You should be able to see the build results under the details because they're made public. Might have to open in-private.

https://syfuhs2.visualstudio.com/Kerberos.NET/_build/results?buildId=987&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76

**Comment by SteveSyfuhs** [comment]
Merged, built, published. Fixes should be available in nuget 4.5.174: https://www.nuget.org/packages/Kerberos.NET/4.5.174

**Comment by filipnavara** [comment]
Thanks a lot!
